### DISCUSSION TITLE
Is anyone working on an IR library?

### DISCUSSION THREAD

**Discussion by Averelll** [toitlang/toit#507] [Drivers]
really enjoying playing around with toit and jaguar, love the live reloading. One of the things holding me back is the lack of libaries for many of the functions I use regularly, one of which is support for ir codes. Anybody working on this, or any idea of how easy it would be to develop such a package?

Thx,
Averell

**Comment by kasperl** [comment]
Hi @Averelll! Welcome to the community. Can you point me towards such an IR library in another language? It typically isn't a lot of work to develop or port such functionality, so if we can find a good starting point with the right license, then we should be able to move quickly on this.

**Reply by nas2011** [reply]
I previously used this: https://github.com/peterhinch/micropython_ir  to get a project rolling in micropython. I had to make some changes to the array sizes to send and receive Daikin AC signals which are much longer than nec or other protocols. Otherwise, these were great libraries to have. 

I'm am super new to both Toit and writing code for anyone other than myself, but I have gone pretty deep in the IR stuff (specifically in micropython in the past) and would be happy to help contribute to a general IR library when the RMT feature is available. If nothing else, I'll put together some stuff for anyone wanting to control their Daikin AC unit.

**Comment by floitsch** [comment]
Also, I believe @lask started work on exposing the ESP32 IR peripheral ("Remote Control RMT").

**Comment by lask** [comment]
That is correct. Please do share useful IR drivers, so we can make sure we support their implementation.

**Reply by nas2011** [reply]
Very excited to have the RMT functionality available. I have a micropython project for an IR ac controller that I want to move to toit and the IR is the only thing I can't implement right now. Any ideas on what the timeline might look like for access to the RMT?

**Reply by lask** [reply]
There is an open PR with the first primitives and a library for RMT https://github.com/toitlang/toit/pull/543

We need to make sure that it can be used for 1-wire (which was our initial motivation for working on it), so the primitives may change a bit. We also want to add a primitive just for receiving.

Is your project open source? If so, I would love to have a look at it to see what we are missing in order to support something like your use case.

**Reply by nas2011** [reply]
I actually just dug this project out and need to clean it up but I'd be happy to share. The basic idea is this:
A big byte string is constructed based on a set of given parameters. The IR pulse duration varies for ones and zeros. There are some headers and spacers in the specific protocol but basically I build the byte string and turn that into a list of pulse durations that are then sent using the RMT function.

<img width="449" alt="image" src="https://user-images.githubusercontent.com/101505861/158492958-7de93d6b-a488-4a74-8f2b-2aceb73161d5.png">

<img width="467" alt="image" src="https://user-images.githubusercontent.com/101505861/158493110-329f979f-fc43-4580-a5ae-f95913193172.png">

An example of the timings might look like this: 
<img width="549" alt="image" src="https://user-images.githubusercontent.com/101505861/158493255-33830c72-649c-4d1e-b967-334c76b9fc13.png">

These are in microseconds based on the clock div. The micropython documentation has a good example of the write pulses feature as well.

<img width="520" alt="image" src="https://user-images.githubusercontent.com/101505861/158493396-08173d08-c0b3-4d12-b9c3-2e4f660a380c.png">

**Reply by lask** [reply]
Thanks for sharing the example (and sorry for the late reply). You should be able to do this in the next release! We have added support for some of the esp32 RMT library, and I think everything you describe should be possible. However, there is a limit to how _big_ a byte string we can handle. The overhead for each bit is pretty significant, so if your string is sufficiently large, then you may still be out of luck.

You can have a sneak peak of the library as it is already committed to master.

**Comment by floitsch** [comment]
The RMT library is now committed.
If you use 2.0.0-alpha (default in the latest Jaguar release), there is access to it.

**Comment by nas2011** [comment]
Thanks for the update! I started playing around but I can't seem to get things to work. I have identical setups running side by side literally same pins same leds and everything, but don't seem to be getting the same output. 

![image](https://user-images.githubusercontent.com/101505861/174680670-931698c4-24dd-4aee-ae64-acda8cd1a451.png)


I know the output isn't the same because one turns on my A/C and the other doesn't. Next step is to break out the receiver and actually look at the pulses being generated but before I go down that rabbit hole, wanted to see if anyone has a thought about why this isn't working from looking at the code. For reference, sigTimes list is a list of the 439 pulse durations needed. 

Thanks

**Reply by floitsch** [reply]
First of all: cool that you got the A/C to react. I always (ab)used the RMT for other purposes. It's good to see that it also works for its intended purpose. :)

Honestly, I don't see why this would work on one ESP32 and not another.

If you have another ESP32 lying around you could use the rmt input functionality to listen for pulses. This way you could see if the pin emits the same pulses.
Edit: grr. 439 pulses are too long for the input buffer. That makes things more difficult :(

side-note: I'm guessing the red lines are, because you have `toit` installed (from toit.io) which doesn't support `rmt` yet?
If you uninstall it, you should get the completion from Jaguar which recognizes the RMT library.

**Reply by floitsch** [reply]
Also note: in theory you don't need 4 memory blocks, as the interrupt should fill the buffer when necessary.
(It shouldn't hurt, though).

**Reply by floitsch** [reply]
Another question: could it be that the 439 pulses are actually way fewer pulses, but that there is a carrier frequency that is applied when the signal is low?
This would explain all these "400" entries. It doesn't really fit with the first 3600, but given that the idle_level is 0, the first 3600 shouldn't be necessary in the first place.

See https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#rmt-transmitter-overview
In the link the carrier is applied when the signal is high, but the esp32 has a functionality to apply it for low as well.

**Reply by robvanlopik** [reply]
Has there been any progress on this topic?

**Reply by floitsch** [reply]
Not that I'm aware of. The RMT library should work (and is used for multiple other libraries, like 1-wire, dht11, or hc-sr04), but I don't think anyone has yet used it for its intended purpose.
